tidb-binlog/pump_client/client.go

```diff
@@ -255,89 +249,129 @@ func (c *PumpsClient) WriteBinlog(binlog *pb.Binlog) error {
-				Logger.Warnf("[pumps client] write commit binlog %d failed, error %v", binlog.CommitTs, err)
-				return nil
+			if time.Since(startTime) > CommitBinlogTimeout {
+				break
 			}
 		} else {
 			if !isRetryableError(err) {
 				// this kind of error is not retryable, return directly.
 				return err
 			}
 
-			// every pump can retry 5 times, if retry 5 times and still failed, set this pump unavaliable, and choose a new pump.
-			if (retryTime+1)%RetryTime == 0 {
-				c.setPumpAvaliable(pump, false)
-				pump = c.Selector.Next(binlog, retryTime/5+1)
-				Logger.Debugf("[pumps client] avaliable pumps: %v, write binlog choose pump %v", c.Pumps.AvaliablePumps, pump)
+			if time.Since(startTime) > c.BinlogWriteTimeout {
+				break
 			}
 
 			retryTime++
-			if retryTime > c.RetryTime {
-				break
-			}
 		}
 
-		time.Sleep(RetryInterval)
+		if binlog.Tp != pb.BinlogType_Prewrite {
+			time.Sleep(RetryInterval * 10)
+		} else {
+			time.Sleep(RetryInterval)
+		}
+	}
+
+	pump, err1 := c.backoffWriteBinlog(req, binlog.Tp)
+	if err1 == nil {
+		return nil
 	}
 
-	return err
+	return errors.Errorf("write binlog failed, the last error %v", err)
+}
+
+func (c *PumpsClient) backoffWriteBinlog(req *pb.WriteBinlogReq, binlogType pb.BinlogType) (pump *PumpStatus, err error) {
+	if binlogType != pb.BinlogType_Prewrite {
+		// never return error for commit/rollback binlog.
+		return nil, nil
+	}
+
+	unAvaliablePumps := make([]*PumpStatus, 0, 3)
+	c.Pumps.RLock()
+	for _, pump := range c.Pumps.UnAvaliablePumps {
+		unAvaliablePumps = append(unAvaliablePumps, pump)
+	}
+	c.Pumps.RUnlock()
+
+	var resp *pb.WriteBinlogResp
+	// send binlog to unavaliable pumps to retry again.
+	for _, pump := range unAvaliablePumps {
+		pump.ResetGrpcClient()
+
+		resp, err = pump.WriteBinlog(req, c.BinlogWriteTimeout)
+		if err == nil {
+			if resp.Errmsg != "" {
+				err = errors.New(resp.Errmsg)
+			} else {
+				// if this pump can write binlog success, set this pump to avaliable.
+				Logger.Debugf("[pumps client] write binlog to unavaliable pump %s success, set this pump to avaliable", pump.NodeID)
+				c.setPumpAvaliable(pump, true)
+				return pump, nil
+			}
+		}
+	}
+
+	return nil, err
+}
+
+func (c *PumpsClient) checkPumpAvaliable() {
+	c.Pumps.RLock()
+	allPumps := copyPumps(c.Pumps.Pumps)
+	c.Pumps.RUnlock()
+
+	for _, pump := range allPumps {
+		if !pump.IsAvaliable() {
+			c.setPumpAvaliable(pump, false)
+		}
+	}
 }
 
 // setPumpAvaliable set pump's isAvaliable, and modify UnAvaliablePumps or AvaliablePumps.
 func (c *PumpsClient) setPumpAvaliable(pump *PumpStatus, avaliable bool) {
-	pump.IsAvaliable = avaliable
-	if pump.IsAvaliable {
-		err := pump.createGrpcClient(c.Security)
-		if err != nil {
-			Logger.Errorf("[pumps client] create grpc client for pump %s failed, error: %v", pump.NodeID, err)
-			pump.IsAvaliable = false
-			return
-		}
+	c.Pumps.Lock()
+	defer c.Pumps.Unlock()
+
+	pump.ResetGrpcClient()
```

Comment:
always call it?

Reply:
if  pump from paused to online, will set pump from unavaliable to available, the old grpc connection may be not available, so need reset;
if pump from available to unavailable, means the grpc connection is unavailable, so need reset.